skbio/metadata/_interval.py

```diff
@@ -0,0 +1,756 @@
+                yield intvl
+
+    def _query_attribute(self, metadata, intervals=None):
+        """Yield ``Interval`` objects based on query attributes.
+
+        Parameters
+        ----------
+        metadata : dict or ``None``
+            If it is ``None``, return empty iterator; if it is
+            ``{}``, return an interator of all the ``Interval``
+            objects.
+        intervals : an iterable of ``Interval`` objects
+        """
+        if metadata is None:
+            return
+
+        if intervals is None:
+            intervals = self._intervals
+
+        for intvl in intervals:
+            for (key, value) in metadata.items():
+                if (key not in intvl.metadata or
+                        intvl.metadata[key] != value):
+                    break
+            else:
+                yield intvl
+
+    @experimental(as_of='0.5.0-dev')
+    @_rebuild_tree
+    def query(self, locations=None, metadata=None):
+        """Yield ``Interval`` object with the locations and attributes.
+
+        The ``Interval`` objects must meet both requirements: 1) overlap
+        with any of the spans specified by ``locations``; 2) satisfy
+        ``metadata`` specification. For instance, you can identify
+        all the recA genes that overlap with (10, 100) or (900, 1000)
+        with this code ``interval_metadata.query([(10, 100),
+        (900, 1000)], {'gene': 'recA'})``.
+
+        Parameters
+        ----------
+        locations : iterable of tuples of int pair, optional
+            Specifies locations to look for the ``Interval``
+            objects. An satisfying interval feature only need to overlap with
+            one location. Default (``None``) means all ``Interval``s meet
+            this requirement.
+
+        metadata : dict, optional
+            A dictionary of key word attributes associated with the
+            ``Interval`` object. It specifies what metadata keywords and
+            values to look for. Default (``None``) means all ``Interval``s
+            meet this requirement.
+
+        Yields
+        ------
+        Interval
+            ``Interval`` object satisfying the search criteria.
+        """
+        if locations is None:
+            for intvl in self._query_attribute(metadata):
+                yield intvl
+        else:
+            for loc in locations:
+                intvls = self._query_interval(loc)
+                if metadata is None:
+                    metadata = {}
+                for intvl in self._query_attribute(metadata, intvls):
+                    yield intvl
+
+    @experimental(as_of='0.5.0-dev')
+    def drop(self, intervals):
+        """Drops Interval objects.
+
+        The given ``Interval`` objects will be removed and their
+        associated ``IntervalMetadata`` will be set to ``None``.
+
+        Parameters
+        ----------
+        interval : iterable of ``Interval``
+            ``Interval`` objects to drop from this object.
+
+        """
```

Comment:
You're right, thank you

On Friday, September 2, 2016, Zech Xu <notifications@github.com> wrote:

> In skbio/metadata/_interval.py
> <https://github.com/biocore/scikit-bio/pull/1414#discussion_r77427471>:
>
> > +                for intvl in self._query_attribute(metadata, intvls):
> > +                    yield intvl
> > +
> > +    @experimental(as_of='0.5.0-dev')
> > +    def drop(self, intervals):
> > +        """Drops Interval objects.
> > +
> > +        The given ``Interval`` objects will be removed and their
> > +        associated ``IntervalMetadata`` will be set to ``None``.
> > +
> > +        Parameters
> > +        ----------
> > +        interval : iterable of ``Interval``
> > +            ``Interval`` objects to drop from this object.
> > +
> > +        """
>
> It is impossible to have a Interval bound to a different IntervalMetadata,
> isn't it? Unless you manipulate the private attr
> IntervalMetadata._intervals directly, in which case there is no way to
> stop users doing damages.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1414/files/1837b1b85f84a93c77811f3c1a425f7da63c4a69#r77427471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwvwMmfCRVfpY0YbAJOeYicQA72vZMhks5qmL-agaJpZM4JoIbD>
> .
>


-- 
Sent from my phone.